### ISSUE TITLE
Expand vertex file reader to parse particle type and momenta

### ISSUE THREAD

**Issue by gipert** [legend-exp/remage#156]
This would make remage support a wide range of external event generators through `RMGVertexFromFile`. The implementation would need to be a bit more complex than what it is right now, since the input file structure is "jagged". I guess a lot of this is already implemented in `RMGGrabmayrGCReader`.

Actually, "vertex" in `RMGVertexFromFile` was originally intended to refer only to vertex position. We should maybe implement all of this as an additional generator.

**Comment by tdixon97** [comment]
I dont think the table necessarily needs to be jagged, we just need to include the evtid (and time?)

**Comment by gipert** [comment]
Yes, I meant that the number of particles originating at the vertex is not fixed.

**Comment by tdixon97** [comment]
Started prototyping a vertex generator here: [pyvertexgen](https://github.com/tdixon97/pyvertexgen). 
It should be easy to write to a HDF5 file and start testing this.

**Comment by ManuelHu** [comment]
yes, but you will need to write it in geant4's ugly HDF5 ntuple format (are there even docs for it?). An LH5 reader would be very hard to add to the geant4 "analysis framework".

**Comment by tdixon97** [comment]
I don't know much about it, maybe we can write a simple converter like is done for the output?

**Comment by tdixon97** [comment]
I had a look and I am not sure how to create the Geant4 Ntuple inputs. Maybe we should go for something like the `RMGGrabmayrGCReader` as @gipert  suggested and forget the geant4 input format.

**Comment by tdixon97** [comment]
@ManuelHu what is the status of this? I recall you had something working

**Comment by ManuelHu** [comment]
not really. I have the general infrastructure to read LH5 files as input in a multithreaded run; but I did not yet work on the particle creation side. Only the vertex (position) reader works at the moment.

And I also still need to push some code I worked on locally, to really decouple the LH5 reading part from the vertex reader. After that, the rest should be rather easy.

**Comment by tdixon97** [comment]
I might be able to find some time to look into creating the vertices if the LH5 reader works. I think it should be fairly straightforward and I'd like to do this relatively soon.

**Comment by ManuelHu** [comment]
@tdixon: see #230 for the refactor. The `RMGAnalysisReader` can only be initialized from the main thread. Apart from that, it should be quite simple to use and adapt.

note: there is no sane way to access/check the structure of the input file with the `G4*AnalysisReader`s. The structure has always to be the same.

**Comment by tdixon97** [comment]
This looks great @ManuelHu , maybe you could add some docstrings to the main functions? That would make it easier for people to use this reader for particular applications, and would be a step towards having proper documentation.

**Comment by tdixon97** [comment]
I think we should also add some test on multithreading / thread safety since that makes me nervous.. 
We could generate some LH5 file of positions, read it in in multithread mode and then compare the output `vertex` table to the input ?

**Comment by ManuelHu** [comment]
Sure. I am also worried quite a lot, the whole situation is very bad:
* the analysis readers automatically mangle names when used on worker threads.
* but we cannot pre-compute input files; as we don't know the number of rows needed
* analysis readers are thread-local _by file type_. there is no way to create a new reader that is not connected to this global instance
* So if we read two HDF5 files in different parts of remage that will be racy, unless we use a global lock around any analysisreader usage.
* ...and we just need to hope that Geant4 does not use it internally
* ouch

**Comment by tdixon97** [comment]
I guess we need to add some checks then?
________________________________

**Comment by ManuelHu** [comment]
I imagine an (ugly) solution. Maybe it works, but I would really test before that if we can actually concurrently read from two files before investing time into making this safe and simple to use.
If it does not work in that simple case (reading `vtx` and `kin` data), we would need to scrap the whole solution and implement the input reading by hand :-/
